File: src/test/java/org/trp/shincolle/entity/base/ShipRecoveryDecisionResolverTest.java
package org.trp.shincolle.entity.base;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ShipRecoveryDecisionResolverTest {

    @Test
    void forcedRecoveryShouldAlwaysBeEligibleForTeleportAttempt() {
        assertTrue(ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                new ShipRecoveryDecisionResolver.State(true, 0.0D, 256.0D)));
    }

    @Test
    void nonForcedRecoveryShouldRequireLongDistance() {
        assertFalse(ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                new ShipRecoveryDecisionResolver.State(false, 256.0D, 256.0D)));
        assertTrue(ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                new ShipRecoveryDecisionResolver.State(false, 256.01D, 256.0D)));
    }

    @Test
    void clearThresholdsShouldUseExclusiveLimits() {
        assertFalse(ShipRecoveryDecisionResolver.shouldClearAfterStuck(120, 120));
        assertTrue(ShipRecoveryDecisionResolver.shouldClearAfterStuck(121, 120));
        assertFalse(ShipRecoveryDecisionResolver.shouldClearAfterMoveFailures(40, 40));
        assertTrue(ShipRecoveryDecisionResolver.shouldClearAfterMoveFailures(41, 40));
    }
}

File: src/test/kotlin/org/trp/shincolle/entity/base/ShipRecoveryDecisionResolverTest.kt
package org.trp.shincolle.entity.base

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ShipRecoveryDecisionResolverTest {

    @Test
    fun `forced recovery should always be eligible for teleport attempt`() {
        assertThat(
            ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                ShipRecoveryDecisionResolver.State(true, 0.0, 256.0)
            )
        ).isTrue()
    }

    @Test
    fun `non forced recovery should require long distance`() {
        assertThat(
            ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                ShipRecoveryDecisionResolver.State(false, 256.0, 256.0)
            )
        ).isFalse()
        assertThat(
            ShipRecoveryDecisionResolver.shouldAttemptTeleport(
                ShipRecoveryDecisionResolver.State(false, 256.01, 256.0)
            )
        ).isTrue()
    }

    @Test
    fun `clear thresholds should use exclusive limits`() {
        assertThat(ShipRecoveryDecisionResolver.shouldClearAfterStuck(120, 120)).isFalse()
        assertThat(ShipRecoveryDecisionResolver.shouldClearAfterStuck(121, 120)).isTrue()
        assertThat(ShipRecoveryDecisionResolver.shouldClearAfterMoveFailures(40, 40)).isFalse()
        assertThat(ShipRecoveryDecisionResolver.shouldClearAfterMoveFailures(41, 40)).isTrue()
    }
}
